### ISSUE TITLE
Getting auth_state to the spawner

### ISSUE THREAD

**Issue by rcthomas** [jupyterhub/jupyterhub#2552]
I think this may be similar to the issue that led to the async `get_options_form`.  Before the home page with named servers is rendered, I want to get some information from an external service so I can customize the home page.  For instance, I want staff to be able to see options related to a testbed system but normal users shouldn't see it, since they can't access it.

I've looked at sticking this on `auth_state` and that seems the right place to put it because it's to do with the user.  But, unless I'm mistaken, to get to it later I need to call `get_auth_state` and that's a coroutine.  I could actually do it some other way, make the call from the post-auth hook, but it all boils down (I think) to needing to await something and get it (or something derived from it) into the template.

The home page `get` is already async, I was wondering if putting something there to possibly await `get_auth_state` and allow it to be used during rendering might be possible.  What I was looking for first was some way to get that information to the spawner in some non-blocking way, but I don't see it.

**Comment by vilhelmen** [comment]
Yeah, I think your best bet is to use `post_auth_hook` to fill out `auth_state` data and then use the callable form of `options_form` and use `self.user.get_auth_state()` to get the auth state. You may need an await in there, I'm extremely bad at async python.

**Comment by rkdarst** [comment]
About this issue and #2555, it seems that the root issue is needing to run async stuff before you render the options form.  I also dynamically adjust available profiles in one of my hubs using kubespawner: I have a profile_list which is a callable, and gets run with the spawner as an argument.  The spawner has the user attribute there.  I do this backwards from you: for each profile, I generate the list of users who have access to it and then comparing it with users is simple, but it could be reversed: you get permissions for each user and do what you do.  I don't do it async (but I should...), but from what I can tell, a callable options_form can run async.  I also do refreshing: after a certain delay, I will refresh all the profiles and their allowed users.

So, to summarize, could you put everything you need into a callable async options_form on the spawner?

I have a feeling you have dug into it so much that you probably would have thought of it if it worked, but just thought to mention it here...

**Comment by rcthomas** [comment]
The problem is `options_form` rendering comes too late in the process.  What I need is to customize the home/console page based on the contents of `auth_state`.  This maybe doesn't make sense unless you are doing named servers, which we use to organize how users can spawn on different systems.

It's true I did have the problem of trying to find some async function I could slot the call into, but right before `render_template` was where I needed it.  It could be there's a more universally proper place to put it, something that gets called and could be used to refresh the auth info automatically.

**Comment by vilhelmen** [comment]
I'm not sure I follow. `post_auth_hook` is part of `get_autheticated_user`, which means it runs before login is completed, which should be well before the spawner for that user is initialized.

Either way, if you're having issues with `options_form`, you should try assigning a function to `profile_list` and loading `auth_state` from there. I believe you'd want to use `auth_state = yield self.user.get_auth_state()` to access it.

**Comment by rcthomas** [comment]
Indeed, I am using `post_auth_hook` to populate `auth_state` but I am _not_ using `options_form` at all.  This is about getting ahold of `auth_state` at a different point than during spawning.

**Comment by vilhelmen** [comment]
I just had a re-read and now it's much more clear. Sorry about that. If you're not against getting your hands a bit dirty and patching the hub code, you might be able to do what you want here: https://github.com/jupyterhub/jupyterhub/blob/2e8ddeb1142a7b17c55863d8262e34bd989bcc2e/jupyterhub/handlers/pages.py#L51

It's even calling some async user functions, so getting auth state should be fine! You may need to tweak the page template to fit your needs.

**Comment by rcthomas** [comment]
#2555 

It's OK.  It's been a long week.  

But I'm glad you agree that looks like the right place!  :)
